docs/docs/tutorials/web_socket_counter.md

```diff
@@ -0,0 +1,410 @@
+  final handler = webSocketHandler(
+    (channel, protocol) {
+      // A new client has connected to our server.
+      print('connected');
+
+      // Send a message to the client.
+      channel.sink.add('hello from the server');
+
+      // Listen for messages from the client.
+      channel.stream.listen(
+        print,
+        // The client has disconnected.
+        onDone: () => print('disconnected'),
+      );
+    },
+  );
+
+  return handler(context);
+}
+```
+
+:::info
+For more information, refer to the [WebSocket documentation](/docs/advanced/web_socket).
+:::
+
+Save the changes and hot reload should kick in ⚡️
+
+Now we should be able to write a simple script to test the WebSocket connection.
+
+## Establishing a WebSocket Connection
+
+Create a new directory called `example` at the project root and create a `pubspec.yaml`:
+
+```yaml
+name: example
+publish_to: none
+
+environment:
+  sdk: '>=2.18.0 <3.0.0'
+
+dependencies:
+  web_socket_channel: ^2.0.0
+```
+
+Next, install the dependencies:
+
+```sh
+dart pub get
+```
+
+Now, create a `main.dart` with the following contents:
+
+```dart
+import 'package:web_socket_channel/web_socket_channel.dart';
+
+void main() {
+  // Connect to the remote WebSocket endpoint.
+  final uri = Uri.parse('ws://localhost:8080/ws');
+  final channel = WebSocketChannel.connect(uri);
+
+  // Listen to messages from the server.
+  channel.stream.listen(print);
+
+  // Send a message to the server.
+  channel.sink.add('hello from the client');
+
+  // Close the connection.
+  channel.sink.close();
+}
+```
+
+We're using [`package:web_socket_channel`](https://pub.dev/packages/web_socket_channel) to connect to our Dart Frog `/ws` endpoint. We can send messages to the server by calling `add` on the `WebSocketChannel` sink. We can listen to incoming messages by subscribing to the `WebSocketChannel` stream.
+
+With the Dart Frog server still running, open a separate terminal, and run the example script:
+
+```bash
+dart example/main.dart
+```
+
+We should see the following output on the client:
+
+```
+hello from the server
+```
+
+On the server we should see the following output:
+
+```
+connected
+hello from the client
+disconnected
+```
+
+Awesome! We've configured a WebSocket handler and established a connection to our server 🎉
+
+## Managing the Counter State
+
+Now that we've configured the WebSocket handler, we're going to shift gears and work on creating a component that will manage the state of the counter.
+
+In this example, we're going to use a cubit from the [Bloc Library](https://bloclibrary.dev) to manage the state of our counter because it provides a reactive API which allows us to stream state changes and query the current state at any given point in time. We're going to use [package:broadcast_bloc](https://pub.dev/packages/broadcast_bloc) which allows blocs or cubits to broadcast their state changes to any subscribed stream channels -- this will come in handy later on.
```

Comment:
```suggestion
In this example, we're going to use a cubit from the [Bloc Library](https://bloclibrary.dev) to manage the state of our counter because it provides a reactive API which allows us to stream state changes and query the current state at any given point in time. We're going to use [package:broadcast_bloc](https://pub.dev/packages/broadcast_bloc) which allows blocs or cubits to broadcast their state changes to any subscribed stream channels — this will come in handy later on.
```